src/toast/tod/crosstalk.py

```diff
@@ -0,0 +1,473 @@
+# py37+
+# from __future__ import annotations
+
+"""Simulate crosstalk between detectors in ToD.
+
+A typical use of this is that in your pipeline script,
+
+1. Add `add_crosstalk_args(parser)` to your argparse parser to add cli args,
+2. Read & create `op_crosstalk` by adding
+
+
+    .. highlight:: python
+    .. code-block:: python
+
+        if args.crosstalk_matrix is not None:
+            op_crosstalk = OpCrosstalk.read(args)
+
+3. Apply the crosstalk matrix to your data after the ToD is prepared (e.g. signal & noise) by adding
+
+
+    .. highlight:: python
+    .. code-block:: python
+
+        if args.crosstalk_matrix is not None:
+            if comm.comm_world is not None:
+                comm.comm_world.barrier()
+            op_crosstalk.exec(data, "tot_signal")
+            if comm.comm_world is not None:
+                comm.comm_world.barrier()
+
+Note that you need to create your crosstalk matri(x|ces)
+in HDF5 container(s) beforehand. A simple class `SimpleCrosstalkMatrix`
+can assist you to write such file,
+with the simple conventions that
+it has the following datasets:
+
+1. names: ASCII names of your detectors, say of length ``n``
+2. data: ``n`` by ``n`` float64 array of your crosstalk matrix ``m``, where you expect
+
+    ToD_crosstalked = m @ ToD_original
+
+    (Assuming row-major as is standard in Python.)
+
+Note that each crosstalk matrix is dense,
+and when multiple matrices are supplied, they are effectively block-diagonal
+(where each given dense matrix is a block.)
+I.e. in order to provide a block-diagonal crosstalk matrix,
+each block (together with the names of the detectors in that block)
+should write to a seperate HDF5 file to avoid unnessary computation.
+"""
+
+import argparse
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+import h5py
+import numpy as np
+
+try:
+    from numba import jit
+except ImportError:
+    jit = None
+
+from ..mpi import get_world
+from ..op import Operator
+from ..utils import Logger
+
+if TYPE_CHECKING:
+    from typing import List  # Optional
+
+    import toast
+
+# py37+
+# COMM: Optional[toast.mpi.Comm]
+# PROCS: int
+# RANK: int
+COMM, PROCS, RANK = get_world()
+logger = Logger.get()
+IS_SERIAL = PROCS == 1
+
+
+def _fma(out: 'np.ndarray[np.float64]', weights: 'np.ndarray[np.float64]', *arrays: 'np.ndarray[np.float64]'):
```

Comment:
On the contrary, it's probably well known in the Python community that even Numpy is quite inefficient, limited by the Python interpreter.

E.g. in the line

```py
out += weight * array
```

Working in pure Python+Numpy, `weight * array` would be created as a new Python (Numpy) object with additional memory use, so by the time it finishes the whole loop, essentially an extra copy the size of the ToD is made, at the mercy of when the Python garbage collector collected them.

This limitation by Python has 2 consequence:

1. more memory use from intermediate Python objects
2. slower as it is allocating new memory, not doing fused-multiply-add directly.

And in our case, the bottlenecks of the crosstalk operator are mainly

1. the MPI communication
2. more memory needed to hold intermediate row sum
3. time to compute the row sum

So this single limitation from Python is affecting 2 out of 3 the main work needed for the crosstalk operator.

One good example in the Python ecosystem to tackle this is [NumExpr](https://numexpr.readthedocs.io/projects/NumExpr3/en/latest/intro.html#expected-performance) where its sole purpose is to avoid creation of intermediate Numpy objects. It is more emphasized in speed than memory use, and they claim a typical speed gain would be 2-4x just by using NumExpr to accelerate them.

Another option would be to write `_fma` in C++ and expose a Python function that we could use, similar to those helpers for allocating memory. I'm not sure if that's feasible given the signature used (`*arrays`.)

Reply:
solved in commit a38d1d6